Python/pystate.c

```diff
@@ -181,8 +181,8 @@ PyInterpreterState_New(void)
         /* overflow or Py_Initialize() not called! */
         PyErr_SetString(PyExc_RuntimeError,
                         "failed to get an interpreter ID");
-        /* XXX deallocate! */
-        interp = NULL;
+        PyMem_RawFree(interp);
```

Comment:
`_PyRuntime.interpreters.head` (and possibly `_PyRuntime.interpreters.main`) must be reverted too.

Alternately, the ID-related code could be moved to before the head/main stuff (i.e. right after the `HEAD_LOCK();` line).

Reply:
I think moving the ID-related code before the `HEAD_LOCK` is more clear.

Reply:
Sorry for the confusion here.  The ID code needs to be protected by the lock, so `HEAD_LOCK()` should move to right above the `if (_PyRuntime.interpreters.next_id < 0) {` line (i.e. move line 184 of the changed file to right before line 174).  Otherwise, moving the head/main stuff down *after* the ID stuff looks good.

Reply:
FWIW, I don't recall why I originally put the ID-related code *after* the the head/main code.  I vaguely remember there was a reason (and wish I had left a comment).  Regardless, at this point I do not see any reason why putting the ID-related code would be a problem. :)

Reply:
Is not going to be a problem if we return in the middle of the `HEAD_LOCK()` - `HEAD_UNLOCK()` macros?

Reply:
In commit b48d5bd I changed the code to return outside the macroblocks.